[3단계 - 질문 삭제하기 리팩터링] 리뷰 요청드립니다.

itdar opened the pull request:
안녕하세요, 리뷰어님.
리뷰요청 드립니다!

질문이 있습니다.
 1. 기존 Question 과 Answer 의 다대일 관계에서 Question 에서 Answer 로의 참조를 위해서 양방향 매핑 확인 후 리팩토링 했습니다. 이 때, 양방향 매핑 하면서 QnAServiceTest 에서 MockitoSetting 관련해서 strictness 를 변경해주거나, 테스트케이스 내용 중 Unnecessary stubbings 가 발견되는 부분을 제거해서 구동하게 해두었는데, (지우지 않고 주석해두었습니다) 맞는지 모르겠습니다.
 2. List<Answer> 을 일급콜렉션으로 래핑하는 과정에서, Question 클래스에서 Answer 의 관계를 맺고 있는 부분 자체를 일급콜렉션으로 바꾸지는 않고 로직이 들어가는 부분에서 래핑해서 동작하게 해두었습니다. 관계 없을까요? 아니면 일급콜렉션을 다대일 양방향에서 사용하도록 만들 수 있을까요?

주말 잘 보내세요. 감사합니다!

itdar committed:
docs: 요구사항 정리 (1단계 - 엔티티 매핑)

itdar committed:
feat: properties 설정

- Spring Data JPA 사용 시 동작 쿼리 로그로 확인 가능하도록
- 어플 실행 시 데이터베이스 테이블 자동 생성하도록 속성 추가

itdar committed:
feat: BaseTimeEntity 클래스 구현 및 어플에 적용

- 다른 entity에서 상속받아 사용될 Auditing entity class
- 어플리케이션 시작부에서 enable auditing 함

itdar committed:
feat: 각각의 entity 클래스들 구현 커밋

- 테스트 실행을 위해 어플 구동이 되게끔 한번에 구현이 필요했음 -> 각 엔티티클래스 동작하여 현재 어플리케이션 정상 구동 되는 상태
- DDL과 정확한 매핑은 아직 확인해서 수정해야함

itdar committed:
test: AnswerTest 추가

- Answer 생성 후, save 테스트
- 어플리케이션 실행하여 각 entity 클래스가 일단 동작하게끔 구현되었나 확인 테스트를 겸함

itdar committed:
test: Answer 테스트케이스 작성

itdar committed:
test: Answer 테스트 추가

itdar committed:
chore: BaseTimeEntity (Auditing class) 속성에 nullable 변경

itdar committed:
test: Question, User 테스트케이스 작성 추가

itdar committed:
refactor:

- Question, User: 불필요 메서드 제거 및 필요 오버라이드 메서드 (equal, hash) 추가
- QnAServiceTest: Quesion과 Answer에서 id 입력받는 생성자 제거했었던 것 복원하며 재수정
- properties에 MySQL dialect로 출력되도록 속성 추가

itdar committed:
fix: DeleteHistory 엔티티클래스의 컬럼 중 enum속성 수정

itdar committed:
test: DeleteHistoryTest 테스트케이스 작성

itdar committed:
fix: 리뷰하신 내용 수정 적용

- @SpringBootTest 어노테이션을 쓰는 것은 테스트 범위가 넓음
  -> @DataJpaTest 어노테이션으로 축소
  -> 이 떄, DataJpaTest 어노테이션을 쓰려면, test database가 기본 embedded db 로 auto configuration 되는 설정을 none 으로 해줘야함 ( @AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE) )

itdar committed:
refactor: 연관관계를 매핑

- 객체를 참조와 DB 테이블의 외래키를 매핑해서 객체에서는 참조를 하고, 테이블에서는 외래키를 사용하도록 연결
- 객체 참조로 매핑하면서 불필요해진 id 들 삭제

```
Answer 일대다 User
Answer 일대다 Question
DeleteHistory 일대다 User
Question 일대다 User
```

TODO: 변경된 구조에 맞게 테스트케이스 통과하도록 해야함

Smallmachines-Gin committed:
docs(README.md): 2단계 - 연관 관계 매핑 요구사항 등 정리

itdar committed:
test: JPA 스터디 내용 별도 패키지로 커밋

itdar committed:
feat: User / Answer / Question / DeleteHistory 각각의 관계 매핑에 따른 엔티티클래스 수정

itdar committed:
test: User / Answer / Question / DeleteHistory 각각의 관계 매핑에 따른 테스트케이스 수정

itdar committed:
Merge branch 'step2' of https://github.com/itdar/jwp-qna into step2

itdar committed:
test: 각 repository의 테스트클래스에, CRUD에 해당하는 테스트케이스 추가 및 수정

itdar committed:
Merge remote-tracking branch 'upstream/itdar' into step2

# Conflicts:
#	README.md
#	src/main/java/qna/domain/Answer.java
#	src/main/java/qna/domain/DeleteHistory.java
#	src/main/java/qna/domain/Question.java
#	src/main/java/qna/domain/User.java
#	src/test/java/qna/domain/AnswerTest.java
#	src/test/java/qna/domain/DeleteHistoryTest.java
#	src/test/java/qna/domain/QuestionTest.java
#	src/test/java/qna/domain/UserTest.java
#	src/test/java/qna/service/QnaServiceTest.java

step1 merge 후 pull한 내용 merge

itdar committed:
Merge remote-tracking branch 'upstream/itdar' into itdar

itdar committed:
docs: 요구사항 등 READ.md 정리 (3단계 - 질문 삭제하기 리팩터링)

itdar committed:
test: Question 클래스에서 Answer 과의 연관관계 변경에 대한 테스트케이스 추가 (다대일에서의 양방향 매핑)

itdar committed:
refactor: QnAService 의 deleteQuestion method 리팩터링

- 서비스 클래스의 메서드 내에서 구현되었던 로직들을 각각의 해당 클래스 (Question, Answer) 내부로 넣음
- Question에서 Answer 들에 대한 참조를 위하여 JPA 설계의 다대일(N:1) 관계에 양방향 매핑 추가
- Answer, DeleteHistory 리스트들 각각 일급콜렉션으로 변경 및 테스트케이스 작성

- QnAServiceTest 에서, 테스트 시 MockitoSettings 관련 이슈로 인해서
  -> @MockitoSettings(strictness = Strictness.LENIENT 추가
  -> 혹은 when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer)); 주석처리 필요함

itdar committed:
refactor: 메서드, 클래스 이름 등 수정

- 클래스명 오탈자 및 메서드명 수정
- Question 에서 delete 후 delete history 이 각각 따로 되어있었으나, 가독성 향상과 의미적으로 같이 붙어다니게 하기 위하여 통합

itdar committed:
style: 로직 라인 순서 변경

wheejuni commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+public class Answers {
+    private List<Answer> answers;
```
질문에 대한 답변은 여기서 드리면 되겠죠?
[이 링크](https://stackoverflow.com/questions/21719164/jpa-embeddable-can-have-elementcollectionembeddable-property)를 참고해서 설계하시면 될 것 같아요.

wheejuni commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,31 @@
+            deleteHistories.add(answer.deleteHistory());
+        }
+
+        return deleteHistories;
```
stream API로 개선 시도해보시죠!

- `.peek()`
- `.collect()`

wheejuni commented on src/main/java/qna/domain/Question.java:
```diff
@@ -47,6 +56,41 @@ public Question(Long id, String title, User writer, String contents) {
+    public DeleteHistories deleteAndHistory(User loginUser) throws CannotDeleteException {
+        validateUserForDelete(loginUser);
+
+        DeleteHistories deleteHistories = new DeleteHistories();
+        Answers answers = new Answers(this.answers);
+
+        deleteHistories.add(this.deleteAndHistory());
+        deleteHistories.addAll(answers.deleteAllAndHistory());
```
아마 일급 컬렉션 객체가 이 클래스의 필드로 적용되지 못해 조금 아쉬운 부분들이 생긴 듯하네요.
그러나 `DeleteHistories` 지역변수는 없앨 수 있을 것 같습니다. 객체를 선언해놓고 하나씩 원소를 추가하는 방식은 사실 일급 콜렉션을 올바르게 사용하는 방향은 아닙니다. 

`answers.deleteAllAndHistory()` 가 호출될때 `DeleteHistories` 가 튀어나오게 할 순 없나요?

wheejuni commented on src/main/java/qna/domain/Question.java:
```diff
@@ -47,6 +56,41 @@ public Question(Long id, String title, User writer, String contents) {
+        }
+    }
+
+    protected void validateAnswersProprietary(User loginUser) throws CannotDeleteException {
+        if (answers.stream().anyMatch(answer -> !answer.isOwner(loginUser))) {
+            throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+        }
+    }
```
이 로직은 `Answers` 에 있어야 할 것 같습니다!

wheejuni commented on src/test/java/qna/service/QnAServiceTest.java:
```diff
@@ -31,25 +34,25 @@
     public void delete_성공() throws Exception {
         when(questionRepository.findByIdAndDeletedFalse(question.getId())).thenReturn(Optional.of(question));
-        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
+//        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
```
기존에 예제 코드에 작성되어 있던 Mockito 관련 세팅이 이제 더 이상 필요없다면, 부담없이 삭제하셔도 됩니다!

wheejuni reviewed (changes requested):
3단계까지 빠르게 진행하셨네요. 수고 많으셨어요. 💯 
코드상에서 수정해보면 좋을 점들도 보이고요, 질문에 대한 제 답변도 달아 두었습니다. 
참고하시고 도움 되었으면 좋겠습니다. 수정 부탁드릴게요. 👍

itdar committed:
refactor: 휘주니님 리뷰 내용 적용

- Answer 들의 소유권 유효성검사 로직 Answers 내부로 이동
- 로컬 변수 할당되는 부분 제거 -> 메서드에서 반환되도록 수정
- foreach 문 -> stream 문으로 수정
- 불필요 주석내용 삭제

itdar committed:
refactor: Answer, Question 의 다대일 양방향 관계에서, Answers 일급콜렉션을 직접 갖도록 리팩터링

- Embedded, Embeddable 어노테이션으로 연결
- 일급콜렉션으로 변경된에 따라 테스트케이스 추가 및 위치 변경
- DeleteHistory 클래스에서 Time 관련 auditing class 상속

itdar committed:
refactor: Question, Answer 에 들어가는 deleted 상태 래핑 클래스 추가 (Deleted)

itdar committed:
refactor: 원시타입 래핑 및 테스트케이스 확인 완료

itdar commented:
안녕하세요, 휘주니님.
다시 리뷰 요청드립니다!

리뷰해주신 내용들로는
Embeddable, Embedded 어노테이션으로 사용 후 일급콜렉션을 속성으로 추가,
로직 위치 이동 (Question에서 Answer 삭제유저 유효성 체크하던 부분을 Answer로 이동),
Question의 삭제메서드 정리,
Answers에서 삭제하는 구문을 스트림으로 변경,
불필요 주석 삭제

추가로 Embedded, Embeddable 어노테이션을 배운김에 원시타입 속성들 중 제약사항 있는 것들을 래핑했습니다.

리뷰 부탁드립니다. 감사합니다. 남은 주말 잘 보내세요!

itdar commented:
> 3단계까지 빠르게 진행하셨네요. 수고 많으셨어요. 💯
> 코드상에서 수정해보면 좋을 점들도 보이고요, 질문에 대한 제 답변도 달아 두었습니다.
> 참고하시고 도움 되었으면 좋겠습니다. 수정 부탁드릴게요. 👍

안녕하세요, 휘주니님.
다시 리뷰 요청드립니다!

리뷰해주신 내용들로는
Embeddable, Embedded 어노테이션으로 사용 후 일급콜렉션을 속성으로 추가,
로직 위치 이동 (Question에서 Answer 삭제유저 유효성 체크하던 부분을 Answer로 이동),
Question의 삭제메서드 정리,
Answers에서 삭제하는 구문을 스트림으로 변경,
불필요 주석 삭제

추가로 Embedded, Embeddable 어노테이션을 배운김에 원시타입 속성들 중 제약사항 있는 것들을 래핑했습니다.

리뷰 부탁드립니다. 감사합니다. 남은 주말 잘 보내세요!

itdar committed:
refactor: lazy fetch type annotation 추가

wheejuni reviewed (approved):
수고 많으셨습니다. 👍 
`@Embeddable` 활용해서 거의 모든 상태값을 래핑하셨네요. 좋습니다. 
다음 미션으로 넘어가주세요~

Smallmachines-Gin commented:
> 수고 많으셨습니다. 👍
> `@Embeddable` 활용해서 거의 모든 상태값을 래핑하셨네요. 좋습니다.
> 다음 미션으로 넘어가주세요~

 리뷰 해주시느라 고생하셨습니다. 감사합니다!
